File: 112. Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if (root == nullptr) return false;
        if (root->left == nullptr && root->right == nullptr)
            if (root->val == sum)
                return true;
                
        bool res = false;
        if (root->left)
            res = res || hasPathSum(root->left, sum - root->val);
        if (root->right)
            res = res || hasPathSum(root->right, sum - root->val);
        
        return res;
    }
    
};

File: python/112. Path Sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        if not root: return False
        if not root.left and not root.right:
            return sum==root.val
        if root.left and root.right:
            return self.hasPathSum(root.left, sum-root.val) or self.hasPathSum(root.right, sum-root.val)
        if root.left:
            return self.hasPathSum(root.left, sum-root.val)
        if root.right:
            return self.hasPathSum(root.right, sum-root.val)
# 思路： dfs，叶子节点判断节点值是否为sum，非叶子节点判断递归判断子节点
# 时间复杂度:O(N), 空间复杂度O(N)           